packages/backend-common/src/reading/GithubUrlReader.ts

```diff
@@ -106,39 +115,120 @@ export class GithubUrlReader implements UrlReader {
       throw new NotModifiedError();
     }
 
+    const { filepath } = parseGitUrl(url);
     const { headers } = await this.deps.credentialsProvider.getCredentials({
       url,
     });
 
+    return this.doReadTree(
+      repoDetails.repo.archive_url,
+      commitSha,
+      filepath,
+      { headers },
+      options,
+    );
+  }
+
+  async search(url: string, options?: SearchOptions): Promise<SearchResponse> {
+    const repoDetails = await this.getRepoDetails(url);
+    const commitSha = repoDetails.branch.commit.sha!;
+
+    if (options?.etag && options.etag === commitSha) {
+      throw new NotModifiedError();
+    }
+
+    const { filepath } = parseGitUrl(url);
+    const { headers } = await this.deps.credentialsProvider.getCredentials({
+      url,
+    });
+
+    const files = await this.doSearch(
+      repoDetails.repo.trees_url,
+      repoDetails.repo.archive_url,
+      commitSha,
+      filepath,
+      { headers },
+    );
+
+    return { files, etag: commitSha };
+  }
+
+  toString() {
+    const { host, token } = this.config;
+    return `github{host=${host},authed=${Boolean(token)}}`;
+  }
+
+  private async doReadTree(
+    archiveUrl: string,
+    sha: string,
+    subpath: string,
+    init: RequestInit,
+    options?: ReadTreeOptions,
+  ): Promise<ReadTreeResponse> {
     // archive_url looks like "https://api.github.com/repos/owner/repo/{archive_format}{/ref}"
     const archive = await this.fetchResponse(
-      repoDetails.repo.archive_url
+      archiveUrl
         .replace('{archive_format}', 'tarball')
-        .replace('{/ref}', `/${commitSha}`),
-      { headers },
+        .replace('{/ref}', `/${sha}`),
+      init,
     );
 
-    const { filepath } = parseGitUrl(url);
     return await this.deps.treeResponseFactory.fromTarArchive({
       // TODO(Rugvip): Underlying implementation of fetch will be node-fetch, we probably want
       //               to stick to using that in exclusively backend code.
       stream: (archive.body as unknown) as Readable,
-      subpath: filepath,
-      etag: commitSha,
+      subpath,
+      etag: sha,
       filter: options?.filter,
     });
   }
 
-  toString() {
-    const { host, token } = this.config;
-    return `github{host=${host},authed=${Boolean(token)}}`;
+  private async doSearch(
+    treesUrl: string,
+    archiveUrl: string,
+    sha: string,
+    query: string,
+    init: RequestInit,
+  ): Promise<ResponseFile[]> {
+    const matcher = new Minimatch(query.replace(/^\/+/, ''));
+
+    // trees_url looks like "https://api.github.com/repos/octocat/Hello-World/git/trees{/sha}"
+    const recursiveTree: GhTreeResponse = await this.fetchJson(
+      treesUrl.replace('{/sha}', `/${sha}?recursive=true`),
+      init,
+    );
+
+    // The simple case is that we got the entire tree in a single operation.
+    if (!recursiveTree.truncated) {
+      const matching = recursiveTree.tree.filter(
```

Comment:
I'm thinking what happens if this request fails for some reason. Should this also throw errors based on the response?

Reply:
Yeah I intended for the error to bubble up. Hm maybe that error won't have the best message then.